### PR TITLE
Adds the Prototype Syndicate Teleporter from Paradise Station

### PR DESCRIPTION
**Pull request by Cenrus**
## About This PR

Adds the prototype syndicate teleporter, a high risk/high reward mobility item for antagonists with an uplink. It can be purchased for 8 TC.

The original PR can be found here: https://github.com/ParadiseSS13/Paradise/pull/14934

I have made some adjustments to make it work on this codebase, please review with caution as I may have made some mistakes doing so. Many thanks to @Qwertytoforty for giving me the go ahead on importing this.

## How it works

* When used, it teleports you 4-8 tiles forward. If the destination turf is a wall, it will attempt an emergency parallel teleport up to 3 tiles. If the emergency teleport fails to find an open tile, the user will be torn to shreds in the wall, gibbing them.

* Has a maximum of 4 charges, using it expends a charge. Charges are automatically recharged every 20 seconds.

* Teleporting into other people (Telefragging) stuns them briefly and deals some light damage. No, this doesn't gib the victim. I'm not THAT cruel.

* When EMPed, the teleporter will activate without an emergency teleport, as a form of counterplay. This should also make teleporting into the AI Core to EMP it less easy as the user is likely to get telefragged in a wall.

* No emergency teleport if the user is in possession of a bag of holding. You can still use this item effectively with a BoH if you have the balls to wear one while using it.

* The uplink item comes in a box with a piece of paper explaining the teleporter in greater detail than the uplink description.

* Unlike the veil shifter used by the cult, this item does not teleport anything you are pulling with you. This prevents nuke ops from teleporting in, stunning the disk holder, and dragging them away.

gmAlYTi4sQ

Have fun and try not to die.

:cl: Qwertytoforty, imported by Cenrus
add: Adds the syndicate teleporter, an experimental teleporter that syndicate agents can buy.
/:cl:

### CODE REVIEW

**Comment by MNarath1** [comment]
hmm has paradise special port conditions too like goon?

**Comment by MNarath1** [comment]
Ah yes just read it up a different Licence

**Comment by utilbird** [comment]
> Ah yes just read it up a different Licence

Everything that's not in paradise's goon subdirectories is AGPL
https://github.com/ParadiseSS13/Paradise/blob/master/COPYING

File: code/game/objects/items/teleportation.dm [hunk 1]
```diff
@@ -234,3 +235,179 @@
...
+	. = ..()
+	. += "<span class='notice'>[src] has [charges] out of [max_charges] charges left.</span>"
+
+/obj/item/teleporter/attack_self(mob/user)
+	attempt_teleport(user, FALSE)
+
+/obj/item/teleporter/process()
+	if(prob(10) && charges < max_charges)
+		charges++
+
```
Comment (utilbird) [inline code/game/objects/items/teleportation.dm]:
```suggestion
	if(charges < max_charges)
		if(prob(10))
			charges++
	else
		STOP_PROCESSING(SSobj, src)
```

File: code/game/objects/items/teleportation.dm [hunk 2]
```diff
@@ -234,3 +235,179 @@
...
+		if(isclosedturf(destination))
+			if(EMP_D == FALSE)
+				var/direction = get_dir(user, destination)
+				panic_teleport(user, destination, direction) //We're in a wall, engage emergency parallel teleport.
+			else
+				get_fragged(user, destination) //EMP teleported you into a wall, you're dead.
+		else
+			telefrag(destination, user)
+			do_teleport(C, destination, channel = TELEPORT_CHANNEL_FREE)
+			charges--
```
Comment (utilbird) [inline code/game/objects/items/teleportation.dm]:
```suggestion
			charges--
			START_PROCESSING(SSobj, src)
```

**Review by utilbird — COMMENTED** [review]
Could optimize this a bit, so it's not always processing

**Comment by bloons3** [comment]
What's the difference between this and the briefcase teleporter?

**Comment by utilbird** [comment]
this is a personal handheld version, it's a direct upgrade for almost all situations with the downside of possible wall gibbing

**Comment by Cenrus** [comment]
The Briefcase Teleporter can teleport anchored structures like consoles, lockers, or bodies around safely while this personal teleporter only works for you.
Pro tip, you can even upgrade the range of your briefcase telepad using a BSRPED and better micro lasers.

File: code/game/objects/items/teleportation.dm [hunk 3]
```diff
@@ -234,3 +235,151 @@
...
+	<br>
+	<b>Warning:</b> Teleporting into walls will activate a failsafe teleport parallel up to 3 meters, but the user will be ripped apart and gibbed in the wall if it fails to find a safe location.<br>
+	<br>
+	Do not expose the teleporter to electromagnetic pulses, or possess a bag of holding while operating it. Unwanted malfunctions may occur.
+"}
+/obj/item/storage/box/syndie_kit/teleporter
+	name = "syndicate teleporter kit"
+
+/obj/item/storage/box/syndie_kit/teleporter/Initialize()
+	..()
```
Comment (park66665) [inline code/game/objects/items/teleportation.dm]:
```suggestion
/obj/item/storage/box/syndie_kit/teleporter/PopulateContents()
```

File: code/game/objects/items/teleportation.dm [hunk 4]
```diff
@@ -234,3 +235,151 @@
...
+	new /obj/item/teleporter(src)
+	new /obj/item/paper/teleporter(src)
+	return
```
Comment (park66665) [inline code/game/objects/items/teleportation.dm]:
```suggestion
```

File: code/game/objects/items/teleportation.dm [hunk 5]
```diff
@@ -234,3 +235,151 @@
...
+		return
+
+	var/teleport_distance = rand(minimum_teleport_distance,maximum_teleport_distance)
+	var/mob/living/carbon/C = user
+	var/turf/mobloc = get_turf(C)
+	var/turf/destination = get_teleport_loc(mobloc,C,teleport_distance,0,0,0,0,0,0)
+	var/list/bagholding = user.GetAllContents(/obj/item/storage/backpack/holding)
+
+	if(isclosedturf(destination))
+		if(EMP_D == FALSE && !(bagholding.len))
```
Comment (park66665) [inline code/game/objects/items/teleportation.dm]:
```suggestion
		if(!EMP_D && !(bagholding.len))
```

File: code/game/objects/items/teleportation.dm [hunk 6]
```diff
@@ -234,3 +235,151 @@
...
+
+/obj/item/teleporter/attack_self(mob/user)
+	attempt_teleport(user, FALSE)
+
+/obj/item/teleporter/process()
+	if(charges < max_charges)
+		if(prob(10))
+			charges++
+	else
+		STOP_PROCESSING(SSobj, src)
```
Comment (park66665) [inline code/game/objects/items/teleportation.dm]:
```suggestion
		return PROCESS_KILL
```

File: code/game/objects/items/teleportation.dm [hunk 7]
```diff
@@ -234,3 +235,151 @@
...
+/obj/item/teleporter/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	return ..()
+
+/obj/item/teleporter/examine(mob/user)
+	. = ..()
+	. += "<span class='notice'>[src] has [charges] out of [max_charges] charges left.</span>"
+
+/obj/item/teleporter/attack_self(mob/user)
+	attempt_teleport(user, FALSE)
```
Comment (park66665) [inline code/game/objects/items/teleportation.dm]:
```suggestion
/obj/item/teleporter/attack_self(mob/user)
	attempt_teleport(user, FALSE)
	..()
```

Not every `attack_self` calls parent, but it is recommended.

File: code/game/objects/items/teleportation.dm [hunk 8]
```diff
@@ -234,3 +235,151 @@
...
+
+/obj/item/teleporter/proc/panic_teleport(mob/user, turf/destination)
+	var/mob/living/carbon/C = user
+	var/turf/mobloc = get_turf(C)
+	var/turf/emergency_destination = get_teleport_loc(destination,C,0,0,1,saving_throw_distance,0,0,0)
+
+	if(emergency_destination)
+		telefrag(emergency_destination, user)
+		do_teleport(C, emergency_destination, channel = TELEPORT_CHANNEL_FREE)
+		charges--
```
Comment (park66665) [inline code/game/objects/items/teleportation.dm:335]:
```suggestion
		charges--
		START_PROCESSING(SSobj, src)
```

File: code/game/objects/items/teleportation.dm [hunk 9]
```diff
@@ -234,3 +235,150 @@
...
+/obj/item/teleporter/attack_self(mob/user)
+	..()
+	attempt_teleport(user, FALSE)
+
+/obj/item/teleporter/process()
+	if(charges < max_charges)
+		if(prob(10))
+			charges++
+	else
+		return PROCESS_KILL
```
Comment (PowerfulBacon) [inline code/game/objects/items/teleportation.dm]: 10% chance to charge every process seems pretty dodgy, would be much better if you instead had a set time between recharges and just used a timer to set the charge, removing the need for processing

**Review by park66665 — APPROVED** [review]
YOLO

**Pull request merged**